### ISSUE TITLE
Problem in Schulze_Proportional_Prefilter in Experimental/Schulze-STV branch

### ISSUE THREAD

**Issue by LiamM32** [julien-boudry/Condorcet#159]
In the Experimental/Schulze-STV branch, try doing Schulze-STV with `Tests/LargeElectionData/TidemanA19.cvotes`. Alternatively, run the test SchulzePrefilterTest. Either way, the prefilter will eliminate candidate _G_, who is supposed to be a winner.

But this seems to be a programming error, not a mathematical problem with what Schulze_Proportional_Prefilter does.

In my debugger, I found that in this election, when going through `Schulze_Proportional_Prefilter::prepareStrongestPaths` (inherited from Schulze_Core), it goes through the second foreach loop more times than it should. For each cycle of the first foreach loop, it should only do the second foreach loop once for every candidate. But for candidate _G_ (key 6), it cycles through each candidate multiple times. Therefore, it runs `$this->StrongestPaths[$candidate_key][$candidate_key_r] = 0;` multiple times for each value of `$candidate_key_r`. I don't know why.

Edit: The above paragraph turned out to be irrelevant to this bug. But still interesting.

When it goes through `Schulze_Proportional_Prefilter::makeStrongestPaths` (inherited from Schulze_Core), it incorrectly determines some direct links which are then set to `$this->strongestPaths`. Even though we can see in the `.cvotes` file that there are some votes that prefer candidate _G_ to others, `$this->strongestPaths[6]` has some values incorrectly set to 0.

Therefore, when it gets to `Schulze_Proportional_Prefilter::makeRanking`, it falsely finds that many candidates are unanimously preferred to _G_.

**Comment by LiamM32** [comment]
I just checked inside `$election->pairwise->pairwise[6]`. These values are also all set to 0.

**Comment by LiamM32** [comment]
It turns out that `Schulze_Core::makeStrongestPaths()` has to be modified. The problem is that it sets a path from one candidate to another to 0 if they have a pairwise loss to the other candidate.

The modification I made & reverted earlier that uses `$M` as an argument would work. However, you may also just remove the condition `if ($election->getPairwise()[$i]['win'][$j] * $M > $election->getPairwise()[$j]['win'][$i])` and it will probably give the same results to the Schulze Method.

**Comment by julien-boudry** [comment]
@LiamM32 Is still an open issue?

**Comment by LiamM32** [comment]
> Is it still an open issue?

Maybe. I haven't solved it. But it might just be a matter of the whole algorithm being flawed. Perhaps we should contact Schulze himself to ask about this.

But I think it's most likely that the algorithm itself is flawed. Based on the pairwise results, it indeed should be eliminating the candidate that it eliminates according to the algorithm I envisioned. But somehow, this candidate 'K' is supposed to be one of two winners in Tideman A22, according to the Schulze program.

Unfortunately, I don't understand Schulze STV enough to understand how K wins in this case.

**Comment by LiamM32** [comment]
I have decided to give up on Schulze_Proportional_Prefilter. I don't know how to make it as aggressive as it can be without ever removing the wrong candidates.

In a [recent commit](https://github.com/julien-boudry/Condorcet/commit/1c46c1f2d217315be8d118dc10c78799d7cee27e), I started remaking Schulze-STV, extended from SingleTransferableVote. While it doesn't output the correct results for all the Tideman tests, the process runs very fast. I feel confident enough that if I ever get it to find the correct results, it will be fast enough that we don't need a prefilter.

No more prefilter. Issue closed.
